### PR TITLE
Add date to document

### LINKED ISSUE #2266: More metadata fields for document; hide Creator Tool
### Description

I want to be able to add more metadata fields via [document](https://typst.app/docs/reference/meta/document/) such as:

- Keywords
- Description
- Rights
- Creator Address
- Creator City
- Creator Postal Code
- Creator Country
- Creator Work Telephone
- Creator Work Email
- Creator Tool (to overwrite `Typst x.x.x`)

### Use Case

Currently, Typst can only set a select few metadata fields. It's also not possible to change/hide the Creator Tool without postprocessing the PDF document.

### PR DESCRIPTION
**Pull request by SekoiaTree**
Fixes #2266.

Possible points of contention:
- Some of these are basically invisible (rating, nickname), and nobody really uses them.
- Some useful metadata properties are absent

Full list of metadata, for convenience:
```
    /// The document's title. This is often rendered as the title of the
    /// PDF viewer window.
    pub title: Option<EcoString>,

    /// The document's authors.
    pub author: Author,

    /// The document's keywords.
    pub keywords: Keywords,

    /// The document's creation date. Requires a positive year, month and day. If any of these aren't given, no date is written.
    pub creation_date: Option<Datetime>,

    /// The document's identifier (a unique set of text strings for this document).
    pub identifier: Identifier,

    /// The document's rating (-1 for rejected, 0 for unrated, 1-5 otherwise).
    /// Rarely used in practice, but usable!
    pub rating: Option<i32>,

    /// The document's nickname.
    pub nickname: Option<EcoString>,

    /// The tool used to create the document. By default, this is your Typst version.
    pub creator_tool: Option<EcoString>,```

### CODE REVIEW

File: crates/typst-library/src/meta/document.rs [hunk 1]
```diff
@@ -106,3 +112,14 @@ cast! {
...
+/// A list of identifiers.
+#[derive(Debug, Default, Clone, Hash)]
+pub struct Identifier(Vec<EcoString>);
+
+cast! {
+    Identifier,
+    self => self.0.into_value(),
+    v: EcoString => Self(vec![v]),
+    v: Array => Self(v.into_iter().map(Value::cast).collect::<StrResult<_>>()?),
+}
```
Comment (laurmaedje) [inline crates/typst-library/src/meta/document.rs:125]: This struct can be removed.

File: crates/typst-library/src/meta/document.rs [hunk 2]
```diff
@@ -30,6 +32,9 @@ pub struct DocumentElem {
     /// The document's keywords.
     pub keywords: Keywords,
 
+    /// The document's creation date. Requires a positive year, month and day. If any of these aren't given, no date is written.
```
Comment (laurmaedje) [inline crates/typst-library/src/meta/document.rs:35]: Could you wrap comments at 80 columns here and elsewhere?

File: crates/typst/src/export/pdf/mod.rs [hunk 3]
```diff
@@ -161,8 +162,8 @@ fn write_catalog(ctx: &mut PdfContext) {
     }
 
     let creator = eco_format!("Typst {}", env!("CARGO_PKG_VERSION"));
-    info.creator(TextStr(&creator));
-    xmp.creator_tool(&creator);
+    info.creator(TextStr(creator.as_str()));
+    xmp.creator_tool(creator.as_str());
```
Comment (laurmaedje) [inline crates/typst/src/export/pdf/mod.rs:166]: why this change?

File: crates/typst-library/src/meta/document.rs (hunk 2 above)
Comment (laurmaedje) [reply crates/typst-library/src/meta/document.rs:35]: I assume the year+month+day requirement is because of xmp-writer's API, right? @reknih will make a new release shortly that allows more fine-grained date writing, so that we can lift it.

File: crates/typst/src/export/pdf/mod.rs [hunk 4]
```diff
@@ -217,7 +233,9 @@ fn write_page_labels(ctx: &mut PdfContext) -> Vec<(NonZeroUsize, Ref)> {
 
     for (i, page) in ctx.pages.iter().enumerate() {
         let nr = NonZeroUsize::new(1 + i).unwrap();
-        let Some(label) = &page.label else { continue };
+        let Some(label) = &page.label else {
+            continue;
+        };
```
Comment (laurmaedje) [inline crates/typst/src/export/pdf/mod.rs:238]: This should be changed back.

File: crates/typst-library/src/meta/document.rs (hunk 2 above)
Comment (reknih) [reply crates/typst-library/src/meta/document.rs:35]: I have released `xmp-writer 0.2.0` which allows writing a date if you have any of the parts, so this can be addressed with an xmp-writer bump. Cross-reference #2374 for a possible implementation.

File: crates/typst/src/export/pdf/mod.rs (hunk 3 above)
Comment (SekoiaTree) [reply crates/typst/src/export/pdf/mod.rs:166]: Accidental from reverting other changes. Removed.

File: crates/typst/src/export/pdf/mod.rs (hunk 4 above)
Comment (SekoiaTree) [reply crates/typst/src/export/pdf/mod.rs:238]: `cargo fmt --check --all` complains about this on my end. I've reverted it anyway, but just so you know.

Comment (SekoiaTree) [reply crates/typst/src/export/pdf/mod.rs]: Yep, CI failed because of that

**Review by reknih — COMMENTED** [review]
Looks good to me now!

File: crates/typst/src/export/pdf/mod.rs (hunk 4 above)
Comment (laurmaedje) [reply crates/typst/src/export/pdf/mod.rs]: Is it maybe because you still have the semicolon behind the continue? That will force a line break in rustfmt. If it's completely reverted it should be fine since CI passes on main.

Comment (SekoiaTree) [reply crates/typst/src/export/pdf/mod.rs]: ah yeah that was it

**Comment by laurmaedje** [comment]
Thanks!
